encodings/alp/src/alp/mod.rs

```diff
@@ -107,42 +106,87 @@ pub trait ALPFloat: private::Sealed + Float + Display + 'static {
-        let mut patch_indices = BufferMut::<u64>::with_capacity(values.len());
-        let mut patch_values = BufferMut::<Self>::with_capacity(values.len());
-        let mut fill_value: Option<Self::ALPInt> = None;
-
-        // this is intentionally branchless
-        // we batch this into 32KB of values at a time to make it more L1 cache friendly
-        let encode_chunk_size: usize = (32 << 10) / size_of::<Self::ALPInt>();
-        for chunk in values.chunks(encode_chunk_size) {
-            encode_chunk_unchecked(
-                chunk,
-                exp,
-                &mut encoded_output,
-                &mut patch_indices,
-                &mut patch_values,
-                &mut fill_value,
-            );
+        exponents: Exponents,
+    ) -> (Buffer<Self::ALPInt>, Buffer<u64>) {
+        let mut encoded = BufferMut::<Self::ALPInt>::with_capacity(values.len());
+        let mut patch_indices = BufferMut::<u64>::empty();
+        for chunk in values.chunks(Self::ENCODE_CHUNK_SIZE) {
+            let (encoded_chunk, patches_indices_chunk) = Self::encode(chunk, exponents);
+            encoded.extend(encoded_chunk);
+            patch_indices.extend(patches_indices_chunk);
+        }
+        (encoded.freeze(), patch_indices.freeze())
+    }
+
+    /// ALP encode the given values using the given exponents.
+    ///
+    /// The index of each value for which encode-decode is not the identity function is returned.
+    ///
+    /// See also: [Self::encode_chunkwise].
+    #[allow(clippy::cast_possible_truncation)] // The patch_indices are known to be valid indices into encoded.
+    fn encode(values: &[Self], exponents: Exponents) -> (Vec<Self::ALPInt>, Vec<u64>) {
+        let (mut encoded, needs_patch): (Vec<Self::ALPInt>, Vec<bool>) = values
+            .iter()
+            .map(|value| {
+                let encoded = unsafe { Self::encode_single_unchecked(*value, exponents) };
+                let maybe_decoded = Self::decode_single(encoded, exponents);
+                let needs_patch = maybe_decoded != *value;
+                (encoded, needs_patch)
+            })
+            .unzip();
+
+        // Patched values either have tiny differences (e.g. 1.01, 1.02, 1.00000001) or big
+        // differences (e.g. 1.01, 1.02, 1000.0). In the latter case, this large value
+        // prevents bitpacking (or forces patches into bitpacking).
+        //
+        // Zero allows bitpacking but prevents frame-of-reference encoding, so we choose the first
+        // successfully encoded value.
+        let patch_indices: Vec<u64> = needs_patch
+            .into_iter()
+            .enumerate()
+            .filter(|(_, needs_patch)| *needs_patch)
+            .map(|(index, _)| index as u64)
+            .collect();
+
+        if !patch_indices.is_empty() {
+            if let Some(fill_value) =
+                Self::first_non_patched_encoded_value(&encoded, &patch_indices)
```

Comment:
just use unwrap_or_default() to fallback to zero

Reply:
the original code also does this: https://github.com/cwida/ALP/blob/main/include/alp/encoder.hpp#L379-L385